File: bindings/ruby/spec/proven_spec.rb
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
# frozen_string_literal: true

# Integration tests for the Proven Ruby FFI binding.
# These tests require libproven.so to be built and available.
# Build with: cd ffi/zig && zig build
#
# Run: bundle exec rspec spec/proven_spec.rb

require "proven"

RSpec.describe Proven do
  it "has version 1.0.0" do
    expect(Proven::VERSION).to eq("1.0.0")
  end

  describe Proven::SafeMath do
    describe ".div" do
      it "divides correctly" do
        expect(described_class.div(10, 2)).to eq(5)
      end

      it "returns nil on division by zero" do
        expect(described_class.div(10, 0)).to be_nil
      end
    end

    describe ".mod" do
      it "computes modulo correctly" do
        expect(described_class.mod(10, 3)).to eq(1)
      end

      it "returns nil on mod by zero" do
        expect(described_class.mod(10, 0)).to be_nil
      end
    end

    describe ".add" do
      it "adds correctly" do
        expect(described_class.add(1, 2)).to eq(3)
      end
    end

    describe ".sub" do
      it "subtracts correctly" do
        expect(described_class.sub(5, 3)).to eq(2)
      end
    end

    describe ".mul" do
      it "multiplies correctly" do
        expect(described_class.mul(3, 4)).to eq(12)
      end
    end
  end

  describe Proven::SafeString do
    describe ".escape_html" do
      it "escapes HTML entities" do
        result = described_class.escape_html("<script>")
        expect(result).to include("&lt;") if result
      end
    end

    describe ".escape_sql" do
      it "escapes single quotes" do
        result = described_class.escape_sql("it's")
        expect(result).to include("''") if result
      end
    end

    describe ".valid_utf8?" do
      it "validates UTF-8 strings" do
        result = described_class.valid_utf8?("hello")
        expect(result).to be(true) if result
      end
    end
  end

  describe Proven::SafePath do
    describe ".traversal?" do
      it "detects traversal sequences" do
        result = described_class.traversal?("../etc/passwd")
        expect(result).to be(true) if result
      end

      it "accepts safe paths" do
        result = described_class.traversal?("normal/path")
        expect(result).to be(false) if result
      end
    end

    describe ".safe?" do
      it "validates safe paths" do
        result = described_class.safe?("safe/path")
        expect(result).to be(true) if result
      end
    end
  end

  describe Proven::SafeEmail do
    describe ".valid?" do
      it "validates email addresses" do
        result = described_class.valid?("user@example.com")
        expect(result).to be(true) if result
      end

      it "rejects invalid emails" do
        result = described_class.valid?("not-an-email")
        expect(result).to be(false) if result
      end
    end
  end

  describe Proven::SafeNetwork do
    describe ".valid_ipv4?" do
      it "validates IPv4 addresses" do
        expect(described_class.valid_ipv4?("192.168.1.1")).to be true
      end

      it "rejects invalid addresses" do
        expect(described_class.valid_ipv4?("invalid")).to be false
      end
    end
  end

  describe Proven::SafeCrypto do
    describe ".constant_time_compare" do
      it "compares equal strings" do
        result = described_class.constant_time_compare("secret", "secret")
        expect(result).to be(true) if result
      end

      it "detects unequal strings" do
        result = described_class.constant_time_compare("secret", "other!")
        expect(result).to be(false) if result
      end
    end

    describe ".random_bytes" do
      it "generates random bytes" do
        result = described_class.random_bytes(16)
        expect(result.bytesize).to eq(16) if result
      end
    end
  end

  describe Proven::SafeFloat do
    describe ".finite?" do
      it "detects finite values" do
        result = described_class.finite?(1.0)
        expect(result).to be(true) if result
      end
    end

    describe ".div" do
      it "divides correctly" do
        result = described_class.div(10.0, 2.0)
        expect(result).to be_within(0.001).of(5.0) if result
      end

      it "returns nil for division by zero" do
        expect(described_class.div(1.0, 0.0)).to be_nil
      end
    end
  end

  describe Proven::SafeAngle do
    describe ".deg_to_rad" do
      it "converts degrees to radians" do
        result = described_class.deg_to_rad(180.0)
        expect(result).to be_within(0.001).of(Math::PI) if result
      end
    end

    describe ".normalize_degrees" do
      it "normalizes negative degrees" do
        result = described_class.normalize_degrees(-90.0)
        expect(result).to be_within(0.001).of(270.0) if result
      end
    end
  end

  describe Proven::SafeJson do
    describe ".valid?" do
      it "validates JSON strings" do
        result = described_class.valid?('{"key": "value"}')
        expect(result).to be(true) if result
      end

      it "rejects invalid JSON" do
        result = described_class.valid?("{invalid}")
        expect(result).to be(false) if result
      end
    end
  end
end

File: bindings/lua/spec/proven_spec.lua
-- SPDX-License-Identifier: Apache-2.0
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Test suite for Proven Lua bindings
--- Run with: busted spec/

local proven = require("proven")
local safe_math = proven.safe_math
local safe_string = proven.safe_string
local safe_path = proven.safe_path
local safe_email = proven.safe_email
local safe_network = proven.safe_network
local safe_crypto = proven.safe_crypto

describe("SafeMath", function()
    it("divides correctly", function()
        assert.are.equal(5, safe_math.div(10, 2))
    end)

    it("returns nil on division by zero", function()
        assert.is_nil(safe_math.div(10, 0))
    end)

    it("computes modulo correctly", function()
        assert.are.equal(1, safe_math.mod(10, 3))
    end)

    it("returns nil on mod by zero", function()
        assert.is_nil(safe_math.mod(10, 0))
    end)

    it("adds correctly", function()
        assert.are.equal(3, safe_math.add(1, 2))
    end)

    it("subtracts correctly", function()
        assert.are.equal(2, safe_math.sub(5, 3))
    end)

    it("multiplies correctly", function()
        assert.are.equal(12, safe_math.mul(3, 4))
    end)
end)

describe("SafeString", function()
    it("escapes HTML entities", function()
        assert.are.equal("&lt;script&gt;", safe_string.escape_html("<script>"))
        assert.are.equal("a &amp; b", safe_string.escape_html("a & b"))
        assert.are.equal("&quot;quoted&quot;", safe_string.escape_html('"quoted"'))
    end)

    it("escapes SQL single quotes", function()
        assert.are.equal("it''s", safe_string.escape_sql("it's"))
    end)

    it("escapes JavaScript special characters", function()
        assert.are.equal("line\\nbreak", safe_string.escape_js("line\nbreak"))
        assert.are.equal("tab\\there", safe_string.escape_js("tab\there"))
    end)

    it("truncates with suffix", function()
        assert.are.equal("he...", safe_string.truncate_safe("hello world", 5))
        assert.are.equal("hi", safe_string.truncate_safe("hi", 10))
    end)
end)

describe("SafePath", function()
    it("detects traversal sequences", function()
        assert.is_true(safe_path.has_traversal("../etc/passwd"))
        assert.is_true(safe_path.has_traversal("~/file"))
        assert.is_false(safe_path.has_traversal("normal/path"))
    end)

    it("validates safe paths", function()
        assert.is_true(safe_path.is_safe("safe/path"))
        assert.is_false(safe_path.is_safe("../unsafe"))
    end)

    it("sanitizes filenames", function()
        assert.are.equal("file__name", safe_path.sanitize_filename("file<>name"))
        assert.are.equal("__secret", safe_path.sanitize_filename("..secret"))
    end)

    it("joins paths safely", function()
        assert.are.equal("/base/a/b", safe_path.safe_join("/base", {"a", "b"}))
        assert.is_nil(safe_path.safe_join("/base", {"../etc"}))
    end)
end)

describe("SafeEmail", function()
    it("validates email addresses", function()
        assert.is_true(safe_email.is_valid("user@example.com"))
        assert.is_false(safe_email.is_valid("not-an-email"))
        assert.is_false(safe_email.is_valid("@invalid.com"))
        assert.is_false(safe_email.is_valid("user@.com"))
    end)

    it("splits email parts", function()
        local parts = safe_email.split("user@example.com")
        assert.are.equal("user", parts.local_part)
        assert.are.equal("example.com", parts.domain)
    end)

    it("normalizes email addresses", function()
        assert.are.equal("User@example.com", safe_email.normalize("User@EXAMPLE.COM"))
    end)
end)

describe("SafeNetwork", function()
    it("validates IPv4 addresses", function()
        assert.is_true(safe_network.is_valid_ipv4("192.168.1.1"))
        assert.is_false(safe_network.is_valid_ipv4("invalid"))
        assert.is_false(safe_network.is_valid_ipv4("256.1.1.1"))
    end)

    it("detects private addresses", function()
        assert.is_true(safe_network.is_private("192.168.1.1"))
        assert.is_true(safe_network.is_private("10.0.0.1"))
        assert.is_true(safe_network.is_private("172.16.0.1"))
        assert.is_false(safe_network.is_private("8.8.8.8"))
    end)

    it("detects loopback addresses", function()
        assert.is_true(safe_network.is_loopback("127.0.0.1"))
        assert.is_false(safe_network.is_loopback("192.168.1.1"))
    end)

    it("detects public addresses", function()
        assert.is_true(safe_network.is_public("8.8.8.8"))
        assert.is_false(safe_network.is_public("192.168.1.1"))
    end)
end)

describe("SafeCrypto", function()
    it("compares strings in constant time", function()
        assert.is_true(safe_crypto.constant_time_compare("secret", "secret"))
        assert.is_false(safe_crypto.constant_time_compare("secret", "other!"))
        assert.is_true(safe_crypto.constant_time_compare("", ""))
    end)

    it("creates zeroed strings", function()
        local zeroed = safe_crypto.secure_zero(4)
        assert.are.equal(4, #zeroed)
        assert.are.equal("\0\0\0\0", zeroed)
    end)
end)
